From 3b9e821f6d789635b2116dc6483a4862f9835124 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Thu, 09 Jun 2022 18:57:03 +0300
Subject: [PATCH] [SW-70527] gaudi2: ras: vdec sram rings repair

This patch added support for BIRA, BISR and BIST operations
for VDEC SRAM rings.

TODO: Add vdec memory init after bist to avoid ecc.

Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
Change-Id: I1f0a3d87a465de3132e6f1179cc70d4f25a550ae
---

diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index 86b6b15..cb8ad1d 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -37,6 +37,7 @@
 #define DELAY_57_USEC			57
 #define DELAY_4K_USEC			4000
 #define DELAY_20_USEC			20
+#define DELAY_104_USEC			104
 
 #define GAUDI2_DFX_NUM			5
 #define GAUDI2_MAX_RING_NUM		28
@@ -412,6 +413,57 @@
 	}
 }
 
+static void gaudi2_vdec_ring_pattern(enum gaudi2_sram_self_repair_op op)
+{
+	uint32_t expected_ones, expected_zeros;
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL8, mmPSOC_DFT_APB_WORD0);
+	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_4K_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_104_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_4K_USEC);
+	} else {
+		preboot_puts("Invalid sram repair operation\n");
+		return;
+	}
+	regmap_writel(GAUDI2_RING_PATTERN_VAL11, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00000000010100000010100000010100;
+	expected_zeros = 0b00000000001011100001011100001011;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+	if (op == SRAM_SELF_REPAIR_OP_BIST)
+		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
+	else
+		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00000000000000000000000000000000;
+	expected_zeros = 0b00000000000000000000000000010000;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+	if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL8, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+
+		regmap_writel(GAUDI2_RING_PATTERN_VAL5, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
+
+		expected_ones = 00000000000000000000000000000000;
+		expected_zeros = 00000000000000000000010010010000;
+		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	}
+}
+
 static void gaudi2_sram_ring_sel(uint32_t dfx_val, uint32_t ring)
 {
 	uint32_t ring_value = BIT(8) | (ring << 3) | 0x5;
@@ -454,7 +506,7 @@
 				[SRAM_RING_TYPE_SOB_ROT] = gaudi2_sob_rot_ring_pattern,
 				[SRAM_RING_TYPE_SRAM] = gaudi2_sram_ring_pattern,
 				[SRAM_RING_TYPE_XBAR_MID] = gaudi2_xbar_mid_ring_pattern,
-				[SRAM_RING_TYPE_VDEC] = NULL,
+				[SRAM_RING_TYPE_VDEC] = gaudi2_vdec_ring_pattern,
 				[SRAM_RING_TYPE_SOB_XDMA] = NULL,
 				[SRAM_RING_TYPE_NIC] = NULL,
 				[SRAM_RING_TYPE_MME] = NULL,
@@ -477,7 +529,8 @@
 			if (ring_type == SRAM_RING_TYPE_SRAM ||
 				ring_type == SRAM_RING_TYPE_TPC ||
 				ring_type == SRAM_RING_TYPE_SOB_ROT ||
-				ring_type == SRAM_RING_TYPE_XBAR_MID) {
+				ring_type == SRAM_RING_TYPE_XBAR_MID ||
+				ring_type == SRAM_RING_TYPE_VDEC) {
 				gaudi2_sram_id_sel();
 				gaudi2_sram_ring_sel(dfx_value_table[dfx], ring + 1);
 				gaudi2_sram_region_ring_pattern[ring_type](op);
@@ -597,6 +650,7 @@
  */
 
 /* TODO: Add Xbar memory init after bist to avoid ecc. */
+/* TODO: Add vdec memory init after bist to avoid ecc. */
 
 static void gaudi2_reset_engines(void)
 {
diff --git a/arch/gaudi2/common/common.h b/arch/gaudi2/common/common.h
index 9f4c2d1..e35e14e 100644
--- a/arch/gaudi2/common/common.h
+++ b/arch/gaudi2/common/common.h
@@ -230,8 +230,10 @@
 #define GAUDI2_RING_PATTERN_VAL5		0x1000013
 #define GAUDI2_RING_PATTERN_VAL6		0x9
 #define GAUDI2_RING_PATTERN_VAL7		0x7
+#define GAUDI2_RING_PATTERN_VAL8		0x1000022
 #define GAUDI2_RING_PATTERN_VAL9		0x1
 #define GAUDI2_RING_PATTERN_VAL10		0x100000d
+#define GAUDI2_RING_PATTERN_VAL11		0x1000025
 #define GAUDI2_RING_PATTERN_VAL12		0x1000008
 #define GAUDI2_RING_PATTERN_VAL13		0x85
 #define GAUDI2_RING_PATTERN_VAL14		0x1000006
